### PR TITLE
add head routes

### LINKED ISSUE #184: Adding .head routes
## 🚀 Feature Proposal

fastify-static could be enhanced by supplying
fastify.head
routes additionally to existing .get.

## Motivation

HTTP HEAD request is a simple way to receive information about length of file, date of modification and etag. see https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD

As it currently stands, every HTTP HEAD request (curl -I) gives 404 error.
The rather new fastify option exposeHeadRoutes does never return when used with fastify-static.

## Example

For files it could be tested whether they need to be downloaded because of updates. Also it might be checked whether there is enough space in local cache.

### PR DESCRIPTION
**Pull request by ckiss**
resolves #184 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by simone-sanfratello — APPROVED** [review]
lgtm

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you please add a test that checks everything still works with the [`exposeHeadRoutes`](https://www.fastify.io/docs/latest/Server/#exposeheadroutes) option of fastify?

Also with `wildcard: true` and `wildcard: false`.

**Comment by ckiss** [comment]
@mcollina I added some more tests as you asked.

File: index.js [hunk 1]
```diff
@@ -92,7 +92,11 @@ async function fastifyStatic (fastify, opts) {
     })
 
     wrap.on('pipe', function () {
-      reply.send(wrap)
+      if (request.method !== 'HEAD') { reply.send(wrap) }
+    })
+
+    wrap.on('finish', function () {
+      if (request.method === 'HEAD') { reply.send() }
```
Comment (mcollina) [inline index.js:99]: I would move this in the event handler above. I think this would not work for big files.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by simoneb** [comment]
@ckiss @mcollina where do we stand with this PR?

**Comment by mcollina** [comment]
@simoneb it's stalled due to https://github.com/fastify/fastify-static/pull/191#discussion_r614826818.

File: index.js (hunk 1 above)
Comment (ckiss) [reply index.js:99]: @mcollina I tried to do this but the `pipe` event is not being triggered when using the HEAD route.

Comment (mcollina) [reply index.js:99]: Oh great!

Comment (mcollina) [reply index.js:99]: Then we can move the `if (request.method === 'HEAD')` out of the two event handler and only add the one that is needed.

Comment (mcollina) [reply index.js:99]: Anyway, I'm landing this now.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**